File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.ts
// function maxProfit(prices: number[]): number {
//     let maxprofit:number=0
//         for(let i:number=0;i<prices.length;i++){
//             for(let j:number=i+1;j<prices.length;j++){
//             if(prices[i]<prices[j])  {  
//                 let profit=prices[j]-prices[i]
//                 if(profit>maxprofit){
//                     maxprofit =profit
//                 }
//                 }
                
//             }
//     }

//     return maxprofit
// }

// function maxProfit(prices: number[]): number {
//     let maxProfit: number = 0;

//     for (let i: number = 0; i < prices.length; i++) {
//         for (let j: number = i + 1; j < prices.length; j++) {
//             if (prices[i] < prices[j]) {
//                 let profit = prices[j] - prices[i];
//                 if (profit > maxProfit) {
//                     maxProfit = profit; // Update the maximum profit if selling on this day yields a higher profit
//                 }
//             }
//         }
//     }

//     return maxProfit;
// }






function maxProfit(prices: number[]): number {
    let maxProfit: number = 0;
    let minPrice: number = Infinity;

    for (let i: number = 0; i < prices.length; i++) {
        if (prices[i] < minPrice) {
            minPrice = prices[i]; // Update the minimum price seen so far
        } else if (prices[i] - minPrice > maxProfit) {
            maxProfit = prices[i] - minPrice; // Update the maximum profit if selling on this day yields a higher profit
        }
    }

    return maxProfit;
}

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # left, right = 0, 1
        # maxProfit = 0

        # while right < len(prices):
        #     if prices[left] < prices[right]:
        #         profit = prices[right] - prices[left]
        #         maxProfit = max(maxProfit, profit)
        #     else: 
        #         left = right
        #     right += 1
        
        # return maxProfit


        mProfit=0
        bPrice=prices[0]
        for i in range(1,len(prices)):
            if bPrice > prices[i]:
                bPrice=prices[i]
            else:
                mProfit=max(mProfit,prices[i]-bPrice)
        return mProfit